File: backend/theta/market/pricing.py
"""Black-Scholes pricer + realized-vol IV estimate. Pure functions, no OpenD.

Used by the backtest to reconstruct option prices from underlying price history
(no free historical option chain exists). IV is estimated from realized vol, which
keeps collected credit conservative (real IV usually exceeds RV).
"""
import datetime as dt
import math

SQRT252 = math.sqrt(252.0)


def norm_cdf(x: float) -> float:
    """Standard normal CDF via erf (no scipy dependency)."""
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _d1_d2(S, K, T, r, sigma):
    v = sigma * math.sqrt(T)
    d1 = (math.log(S / K) + (r + 0.5 * sigma * sigma) * T) / v
    return d1, d1 - v


def bs_price(S: float, K: float, T: float, r: float, sigma: float, opt: str) -> float:
    """Black-Scholes price. T in years. opt = 'call' | 'put'. T<=0 or sigma<=0 -> intrinsic."""
    opt = opt.lower()
    if T <= 0 or sigma <= 0:
        return max(S - K, 0.0) if opt == "call" else max(K - S, 0.0)
    d1, d2 = _d1_d2(S, K, T, r, sigma)
    disc = math.exp(-r * T)
    if opt == "call":
        return S * norm_cdf(d1) - K * disc * norm_cdf(d2)
    return K * disc * norm_cdf(-d2) - S * norm_cdf(-d1)


def bs_delta(S: float, K: float, T: float, r: float, sigma: float, opt: str) -> float:
    """Option delta. Call in [0,1], put in [-1,0]."""
    opt = opt.lower()
    if T <= 0 or sigma <= 0:
        if opt == "call":
            return 1.0 if S > K else 0.0
        return -1.0 if S < K else 0.0
    d1, _ = _d1_d2(S, K, T, r, sigma)
    return norm_cdf(d1) if opt == "call" else norm_cdf(d1) - 1.0


def estimate_iv(closes, window: int = 20, factor: float = 1.15) -> float:
    """Annualized realized vol of the last `window` log returns, scaled by `factor`.

    Returns 0.0 if there is not enough data. factor approximates the IV premium over RV.
    """
    if closes is None or len(closes) < 2:
        return 0.0
    seg = list(closes)[-(window + 1):]
    rets = [math.log(seg[i] / seg[i - 1]) for i in range(1, len(seg))]
    n = len(rets)
    if n < 2:
        return 0.0
    mean = sum(rets) / n
    var = sum((x - mean) ** 2 for x in rets) / (n - 1)  # sample std, ddof=1
    return math.sqrt(var) * SQRT252 * factor


def strike_for_delta(S, T, r, sigma, target_delta: float, opt: str, step: float = 1.0):
    """Grid-search the OTM strike whose |delta| is closest to target_delta.

    put  -> strikes at/below spot; call -> strikes at/above spot.
    """
    opt = opt.lower()
    target = abs(target_delta)
    lo, hi = S * 0.5, S * 1.5
    k = math.floor(lo / step) * step
    best_k, best_err = None, float("inf")
    while k <= hi:
        if k > 0 and ((opt == "put" and k <= S) or (opt == "call" and k >= S)):
            err = abs(abs(bs_delta(S, k, T, r, sigma, opt)) - target)
            if err < best_err:
                best_err, best_k = err, k
        k += step
    return best_k


def spread_value(spot: float, short_strike: float, long_strike: float,
                 T: float, r: float, sigma: float, direction: str) -> float:
    """Cost per share to close a credit spread right now. Always in [0, width].

    This is the honest mark: at T>0 an out-of-the-money spread still costs
    something to buy back, because the short leg has time value left. Pricing it
    at expiration intrinsic instead (the old behaviour) reports the whole credit
    as profit while the trade is still live.

    direction: "SELL_PUT" (bull put) or "SELL_CALL" (bear call).
    """
    opt = "put" if direction == "SELL_PUT" else "call"
    short_leg = bs_price(spot, short_strike, T, r, sigma, opt)
    long_leg = bs_price(spot, long_strike, T, r, sigma, opt)
    width = abs(short_strike - long_strike)
    return min(max(short_leg - long_leg, 0.0), width)


def years_to_expiry(expiry: str, today: dt.date) -> float:
    """Calendar days from `today` to ISO date `expiry`, in years. Floored at 0."""
    exp = dt.date.fromisoformat(expiry)
    return max((exp - today).days, 0) / 365.0

File: frontend/src/lib/demo/pricing.ts
// Black-Scholes, ported from theta/market/pricing.py and payoff.py, so the demo's payoff surface is
// priced the same way the real desk prices one. Pure.
import type { Direction, PayoffGrid } from "../api";

const R = 0.04;

/** Standard normal CDF (Abramowitz–Stegun 7.1.26 erf; well inside a cent at these sizes). */
export function normCdf(x: number): number {
  const z = Math.abs(x) / Math.SQRT2;
  const t = 1 / (1 + 0.3275911 * z);
  const erf = 1 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t + 0.254829592) * t * Math.exp(-z * z);
  return 0.5 * (1 + (x >= 0 ? erf : -erf));
}

export function bsPrice(S: number, K: number, T: number, sigma: number, opt: "call" | "put"): number {
  if (T <= 0 || sigma <= 0) return opt === "call" ? Math.max(S - K, 0) : Math.max(K - S, 0);
  const v = sigma * Math.sqrt(T);
  const d1 = (Math.log(S / K) + (R + 0.5 * sigma * sigma) * T) / v;
  const d2 = d1 - v;
  const disc = Math.exp(-R * T);
  return opt === "call" ? S * normCdf(d1) - K * disc * normCdf(d2) : K * disc * normCdf(-d2) - S * normCdf(-d1);
}

export function bsDelta(S: number, K: number, T: number, sigma: number, opt: "call" | "put"): number {
  if (T <= 0 || sigma <= 0) return opt === "call" ? (S > K ? 1 : 0) : (S < K ? -1 : 0);
  const d1 = (Math.log(S / K) + (R + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
  return opt === "call" ? normCdf(d1) : normCdf(d1) - 1;
}

/** Cost per share to close a credit spread now. Always in [0, width]. */
export function spreadValue(spot: number, short: number, long: number, T: number, sigma: number, direction: Direction): number {
  const opt = direction === "SELL_PUT" ? "put" : "call";
  const v = bsPrice(spot, short, T, sigma, opt) - bsPrice(spot, long, T, sigma, opt);
  return Math.min(Math.max(v, 0), Math.abs(short - long));
}

const MAX_DAY_ROWS = 31;

function dayAxis(daysLeft: number): number[] {
  const d = Math.max(Math.trunc(daysLeft), 0);
  if (d < MAX_DAY_ROWS) return Array.from({ length: d + 1 }, (_, i) => d - i);
  const n = MAX_DAY_ROWS - 1;
  return [...new Set(Array.from({ length: n + 1 }, (_, i) => Math.round((d * i) / n)))].sort((a, b) => b - a);
}

const r2 = (x: number) => Math.round(x * 100) / 100;

export function payoffGrid(direction: Direction, short: number, long: number, credit: number, contracts: number,
  spot: number, daysLeft: number, sigma: number, nSpots = 41): PayoffGrid {
  const width = Math.abs(short - long);
  const lo = Math.min(spot, short, long) * 0.9;
  const hi = Math.max(spot, short, long) * 1.1;
  const step = (hi - lo) / (nSpots - 1);
  const spots = Array.from({ length: nSpots }, (_, i) => Math.round((lo + i * step) * 1e4) / 1e4);
  const days = dayAxis(daysLeft);
  const pnl = days.map((d) => spots.map((s) => r2((credit - spreadValue(s, short, long, d / 365, sigma, direction)) * 100 * contracts)));
  const breakeven = direction === "SELL_PUT" ? short - credit : short + credit;
  return {
    spots, days, pnl, breakeven: Math.round(breakeven * 1e4) / 1e4,
    max_profit: r2(credit * 100 * contracts), max_loss: r2(-(width - credit) * 100 * contracts), spot, sigma,
  };
}
